File: src/leetcode/python/TrappingRainWater_42.py
class TrappingRainWater_42(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        if height > 1:
            head, tail = 0, len(height) - 1
            water = 0
            while head < tail:
                if height[head] < height[tail]:
                    pt = head + 1
                    while pt <= tail and height[pt] < height[head]:
                        water += height[head] - height[pt]
                        pt += 1
                    head = pt
                else:
                    pt = tail - 1
                    while pt >= head and height[pt] < height[tail]:
                        water += height[tail] - height[pt]
                        pt -= 1
                    tail = pt
            return water
        return 0

File: src/leetcode/java/TrappingRainWater_42.java
package leetcode.java.hard;

public class TrappingRainWater_42 {
  public int trap(int[] height) {
    if (height.length > 1) {
      int head = 0;
      int tail = height.length - 1;
      int pt;
      int water = 0;
      while (head < tail) {
        if (height[head] <= height[tail]) {
          pt = head + 1;
          while (pt <= tail && height[pt] < height[head]) {
            water += height[head] - height[pt];
            pt++;
          }
          head = pt;
        } else {
          pt = tail - 1;
          while (pt >= head && height[pt] < height[tail]) {
            water += height[tail] - height[pt];
            pt--;
          }
          tail = pt;
        }
      }
      return water;
    }
    return 0;
  }
}
